timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/LockCollection.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.lock;
+
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Lists;
+import com.palantir.lock.LockDescriptor;
+
+public class LockCollection {
+
+    private final LoadingCache<LockDescriptor, ExclusiveLock> locksById = CacheBuilder.newBuilder()
+            .weakValues()
```

Comment:
Yes, it can lead to multiple locks existing for the same descriptor, but not at the same time. The first one would have to be deallocated before the second can be created.